Translate 1-js/09-classes/02-class-inheritance (#366)
javascript-tutorial/zh.javascript.info: 1-js/09-classes/02-class-inheritance/article.md

rabbit.hide(); // White Rabbit hides!
```

现在 `Rabbit` 代码变简洁了一点，因为它默认以 `Animal` 作为其构造函数，同时它仍然能良好“运行”，就像 animals 一样。

在其内部，`extends` 关键字将 `[[Prototype]]` 引用从 `Rabbit.prototype` 添加到 `Animal.prototype`：

![](animal-rabbit-extends.png)

因此，如果在 `Rabbit.prototype` 中没有找到某个方法，JavaScript 将会从 `Animal.prototype` 中获取它。

> 我们可以回忆一下这一章 <info:native-prototypes>，JavaScript 对内置对象使用相同的类型继承。例如，`Date.prototype.[[Prototype]]` 是 `Object.prototype`，所以 dates 有通用的对象方法。

Reviewer: `JavaScript 内置对象同样也是基于/使用原型继承的。`
这句修改成这样是不是稍微易读一点？

> 我们可以回忆一下这一章 <info:native-prototypes>，JavaScript 对内置对象使用相同的类型继承。例如，`Date.prototype.[[Prototype]]` 是 `Object.prototype`，所以 dates 有通用的对象方法。

Reviewer 2: OK没问题

> 我们可以回忆一下这一章 <info:native-prototypes>，JavaScript 对内置对象使用相同的类型继承。例如， `Date.prototype.[[Prototype]]` 是 `Object.prototype`，所以 dates 有通用的对象方法。

Reviewer: ```suggestion
我们可以回忆一下这一章 <info:native-prototypes>，JavaScript 对内置对象使用相同的类型继承。例如，`Date.prototype.[[Prototype]]` 是 `Object.prototype`，所以 dates 有通用的对象方法。
```

}

*!*
// 不生效！

> let rabbit = new Rabbit("White Rabbit", 10); // Error: this 未定义

Reviewer: ```suggestion
let rabbit = new Rabbit("White Rabbit", 10); // Error: Error: this is not defined.
```

}

*!*
// 现在可以了
let rabbit = new Rabbit("White Rabbit", 10);
alert(rabbit.name); // White Rabbit
alert(rabbit.earLength); // 10
*/!*
```

> ## Super：内部基于 [[HomeObject]] 实现

Reviewer 2: [Super：内部基于 [[HomeObject]] 实现]  -> [Super 内部探究，[[HomeObject]]]

*/!*
```

代码无法再运行了！我们可以看到，在试图调用 `longEar.eat()` 时抛出了错误。

原因可能不那么明显，但是如果我们跟踪 `longEar.eat()` 的调用，就可以发现原因。在 (\*) 和 (\*\*) 这两行中，`this` 的值都是当前对象（`longEar`）。这是至关重要的一点：所有的对象方法都将当前对象作为 `this`，而非原型或其他什么东西。

因此，在 (\*) 和 (\*\*) 这两行中，`this.__proto__` 的值是完全相同的：都是 `rabbit`。在这个无限循环中，他们都调用了 `rabbit.eat`，而不是在原型链上向上寻找方法。

这张图介绍了发生的情况：

![](this-super-loop.png)

1. 在 `longEar.eat()` 中，(\*\*) 这一行调用 `rabbit.eat` 并且此时 `this=longEar`。
    ```js
    // 在 longEar.eat() 中 this 指向 longEar
    this.__proto__.eat.call(this) // (**)
    // 变成了
    longEar.__proto__.eat.call(this)
    // 即等同于
    rabbit.eat.call(this);
    ```
2. 之后在 `rabbit.eat` 的 (\*) 行中，我们希望将函数调用在原型链上向更高层传递，但是因为 `this=longEar`，因此 `this.__proto__.eat` 又是 `rabbit.eat`！

    ```js
    // 在 rabbit.eat() 中 this 依旧等于 longEar
    this.__proto__.eat.call(this) // (*)
    // 变成了
    longEar.__proto__.eat.call(this)
    // 再次等同于
    rabbit.eat.call(this);
    ```

> 3. ...所以 `rabbit.eat` 不停地循环调用自己，因此它无法进一步地往原型链的更高层调用

Reviewer: ```suggestion
3. ...所以 `rabbit.eat` 不停地循环调用自己，因此它无法进一步地往原型链的更高层调用。
```

};

*!*
// 正常运行
longEar.eat();  // Long Ear eats.
*/!*
```

> 它按照预期运行，基于 `[[HomeObject]]` 技巧。像 `longEar.eat` 这样的方法，知道 `[[HomeObject]]`，并且从它的原型中获取父类方法。并没有使用 `this`。

Reviewer 2: [技巧] -> [运行机制]

### 方法并不是 “自由” 的

> 在前面我们已经知道，通常函数都是 “自由” 的，并没有绑定到 JavaScript。正因如此，它们可以在对象之间复制，并且用另外一个 `this` 调用它。

Reviewer 2: [并没有绑定到 JavaScript] -> [并没有绑定到 JavaScript 中的对象]

> `[[HomeObject]]` 的存在违反了这个原则，因为方法记住了它们的对象。`[[HomeObject]]` 不能被修改，所以这个绑定时永久的。

Reviewer: ```suggestion
`[[HomeObject]]` 的存在违反了这个原则，因为方法记住了它们的对象。`[[HomeObject]]` 不能被修改，所以这个绑定是永久的。
```

*/!*
```

> 调用 `tree.sayHi()` 显示 "I'm an animal"。绝对是错误的。

Reviewer: ```suggestion
调用 `tree.sayHi()` 显示 “I'm an animal”。绝对是错误的。
```

原因很简单：
- 在 (\*) 行，`tree.sayHi` 方法从 `rabbit` 复制而来。也许我们只是想避免重复代码？
- 所以它的 `[[HomeObject]]` 是 `rabbit`，因为他是在 `rabbit` 中创建的。没有办法修改 `[[HomeObject]]`。
- `tree.sayHi()` 内具有 `super.sayHi()`。它从 `rabbit` 中上溯，然后从 `animal` 中获取方法。

![](super-homeobject-wrong.png)

### 方法，不是函数属性

`[[HomeObject]]` 是为类和普通对象中的方法定义的。但是对于对象来说，方法必须确切指定为 `method()`，而不是 `"method: function()"`。

这个差别对我们来说可能不重要，但是对 JavaScript 来说却是非常重要的。

下面的例子中，使用非方法（non-method）语句进行比较。`[[HomeObject]]` 属性未设置，并且继承不起作用：

```js run
let animal = {

>   eat: function() { // 可以用短语法来书写：eat() {...}

Reviewer 2: [可以用短语法来书写] -> [可以使用简短写法]

};

*!*
rabbit.eat();  // 错误调用 super（因为这里并没有 [[HomeObject]]）
*/!*
```

## 总结

1. 扩展类：`class Child extends Parent`：
    - 这就意味着 `Child.prototype.__proto__` 将是 `Parent.prototype`，所以方法被继承。
2. 重写构造函数：
    - 在使用 `this` 之前，我们必须在 `Child` 构造函数中将父构造函数调用为 `super()`。
3. 重写方法：
    - 我们可以在 `Child` 方法中使用 `super.method()` 来调用 `Parent` 方法。
4. 内部工作：
    - 方法在内部 `[[HomeObject]]` 属性中记住它们的类/对象。这就是 `super` 如何解析父类方法的。
    - 因此，将一个带有 `super` 的方法从一个对象复制到另一个对象是不安全的。

补充：

> - 箭头函数没有自己的 `this` 或 `super`，所以它们能很好的适应周围的环境。

Reviewer 2: [所以它们能很好的适应周围的环境] -> [所以它们能融入到就近的上下文，像透明似的]
